BUG: distutils: use correct top-level package name

xoviat opened the pull request:
Closes gh-10219.

letmaik commented on numpy/distutils/command/build_ext.py:
```diff
@@ -262,15 +262,17 @@ def run(self):
-                os.makedirs(runtime_lib_dir)
-            runtime_lib = os.path.join(self.extra_dll_dir, fn)
-            copy_file(runtime_lib, runtime_lib_dir)
+        root_pkg = self.distribution.root_pkg
```
In which cases would using `root_pkg` not work?

xoviat replied:
Where there isn't a `__config__.py`, which is mainly the f2py tests.

letmaik replied:
How is a root package defined? Can one distribution have multiple root packages?

xoviat replied:
Technically, it can, in which case this scheme would fail. I need some real-world examples in order to figure out how to handle them though. This fixes the example that you mentioned. If you have other examples, please share them.

I'm also going to hint that this is a private attribute because I don't want people relying on this.

xoviat committed:
MAINT: add .cache to gitignore

xoviat commented:
Actually, it would be nice to align this with auditwheel/delocate.

@njsmith How does auditwheel decide where to put the libs? If there are multiple packages? If the extension modules aren't in a package, but are put in site-packages directly?

xoviat commented:
Update: auditwheel appears to disambiguate packages.

xoviat commented:
That avoids any magic, aligns the shared lib name with auditwheel, and should be highly robust.

pv commented on numpy/distutils/command/build_ext.py:
```diff
@@ -262,15 +262,24 @@ def run(self):
+        # resulting in a larger wheel than is required. This should be fixed,
+        # but it's so rare that I won't bother to handle it.
+        pkg_roots = set(
+            self.get_ext_fullname(ext.name).split('.')[0]
```
Does it do the right thing with namespace packages?

xoviat replied:
I'm using pretty much [exactly the same logic](https://github.com/pypa/auditwheel/blob/5b8b569d84d21153877d16aaa6786ddd2e0a0411/auditwheel/repair.py#L55) as auditwheel. In the case of a namespace package, `__config__` **should** locate the `extra-dll-dir` as relative to the package root (though it doesn't currently nor with this PR). Is `package.subpackage.__config__` a thing?

xoviat commented on numpy/distutils/command/build_ext.py:
```diff
@@ -262,15 +262,24 @@ def run(self):
+        for pkg_root in pkg_roots:
+            shared_lib_dir = os.path.join(self.build_lib, pkg_root,
+                                          '.libs')
+            if not os.path.isdir(shared_lib_dir):
```
Note: if installing in editable mode, the `shared_lib_dir` should NOT be prefixed with `build_lib` if not using 2to3 (why are there so many different scenarios to handle?). Im just going to leave the latter scenario out for now and handle the former.

xoviat committed:
BUG: distutils: use correct top-level package name

Here, we align the extra-dll name to the one
that auditwheel uses, and we handle the case where
1) there is more than one root package and
2) the root package name is different
than the distribution name

xoviat committed:
CI: appveyor: test OpenBLAS with editable mode

xoviat commented:
Summary of recent changes: fixed editable mode with OpenBLAS as well.

letmaik commented on numpy/distutils/command/build_ext.py:
```diff
@@ -120,7 +120,7 @@ def run(self):
 
         # Setup directory for storing generated extra DLL files on Windows
-        self.extra_dll_dir = os.path.join(self.build_temp, 'extra-dll')
+        self.extra_dll_dir = os.path.join(self.build_temp, '.libs')
```
Is it necessary to change the folder name as part of this fix? I know of at least one library that has `extra-dll` hard-coded in its own `__config__.py`-like script. Keeping the old name would not break those cases.

xoviat replied:
Sorry for not getting this right the first time, but this should have been named `.libs` to begin with as  auditwheel does because it serves the same function.

charris replied:
If this change goes in, it needs a compatibility entry in the release notes.

charris replied:
@letmaik What library is that? It worries me that such things are out there.

charris replied:
This section seems new in 1.14, so there is still time time to change it. @pv thoughts?

pv replied:
As far as scipy is concerned, this doesn't matter and does not introduce extra work. I'm not sure if trying to match auditwheel is a strong argument as the name is quite arbitrary (as long as it is not a valid name for a subpackage), but maybe using same name makes it easier for downstream repackagers (pyinstaller?) etc.

xoviat replied:
This is an implementation detail that should not be relied on because `__config__.py` is the public API. Perhaps this should be clarified in the docs but this was not done in SciPy and it therefore has no effect on SciPy.

xoviat replied:
Thinking long term it may be possible to push this into Python itself but that is not possible currently because extra-dll is a valid package name.

letmaik replied:
There seems to be quite a bit out there: https://github.com/search?q=extra-dll&type=Code&utf8=%E2%9C%93. Ignore the `NCAR/wrf-python` repo, this was a fix brought in by myself and I can change it easily if necessary.

xoviat replied:
Doesn't matter because this was not yet released. We  cannot guarantee API stability in the development branch -- that's unreasonable.

letmaik replied:
OK nevermind, didn't realize this was a new thing, go for it then. :)

xoviat commented:
@charris Would you consider this for a backport? 1.14 isn't even released and this is a bugfix. If it goes to there should be no compatibility concerns

xoviat commented:
Note: minimal changes extracted to gh-10272.

charris commented:
OK, lets try this. I don't like putting it in after the first rc, but better now than later. Thanks @xoviat .

xoviat commented:
@charris This is for 1.15. gh-10272 was extracted for 1.14. This patch should not go into 1.14 because it hasn't been in master long enough.

charris commented:
@xoviat . If this is something folks need to know about, could you put together something for the 1.15 release notes?

xoviat commented:
This is not something that folks need to know about (for 1.15) if they just use 1.14. The issue was that previously people were using unreleased versions of NumPy because this functionality wasn't yet released.

xoviat commented:
I hope that makes sense.
